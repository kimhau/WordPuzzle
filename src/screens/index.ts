export { default as Congratz } from './Congratz/Congratz';
export { default as Game } from './Game/Game';
export { default as Home } from './Home/Home';
export { default as Leaderboard } from './Leaderboard/Leaderboard';
export { default as Startup } from './Startup/Startup';
