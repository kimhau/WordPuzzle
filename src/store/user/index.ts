import { createSlice } from '@reduxjs/toolkit';

const slice = createSlice({
  name: 'user',
  initialState: { name: null, points: 0 } as UserState,
  reducers: {
    updateUser: (state, { payload: { name, points } }: UserPayload) => {
      if (typeof name !== 'undefined') {
        state.name = name;
      }
      if (typeof points !== 'undefined') {
        state.points = points;
      }
    },
    updatePoints: (state, { payload: { points } }: UserPayload) => {
      if (typeof points !== 'undefined') {
        state.points += points;
      }
    },
  },
});

export const { updateUser, updatePoints } = slice.actions;

export default slice.reducer;

export type UserState = {
  name: string | null;
  points: number;
};

export type UserPayload = {
  payload: Partial<UserState>;
};
