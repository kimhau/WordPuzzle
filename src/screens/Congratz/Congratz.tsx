import React, { useCallback, useEffect, useRef, useState } from 'react';
import { useTranslation } from 'react-i18next';
import {
  Modal,
  Pressable,
  SafeAreaView,
  Text,
  TextInput,
  View,
} from 'react-native';
import { TouchableOpacity } from 'react-native-gesture-handler';
import Share from 'react-native-share';
import ViewShot from 'react-native-view-shot';
import { useDispatch, useSelector } from 'react-redux';
import { CongratzProps } from 'WordPuzzle/@types/navigation';
import { useTheme } from 'WordPuzzle/src/hooks';
import { updatePoints, updateUser, UserState } from 'WordPuzzle/src/store/user';

const Congratz = ({ route, navigation }: CongratzProps) => {
  const dispatch = useDispatch();
  const [modalVisible, setModalVisible] = useState(false);
  const user = useSelector((state: { user: UserState }) => state.user);
  const [name, setName] = useState(user.name);
  const { points, category } = route.params;
  const { Layout, Fonts, Common, Gutters } = useTheme();
  const nameInputRef = useRef<TextInput>(null);
  const { t } = useTranslation('congratz');
  const viewShotRef = useRef<ViewShot>(null);

  useEffect(() => {
    dispatch(updatePoints({ points }));
  }, []);

  const onPressNext = useCallback(() => {
    if (name) {
      navigation.reset({
        index: 1,
        routes: [
          {
            name: 'Home',
          },
          {
            name: 'Game',
            params: { category },
          },
        ],
      });
    } else {
      setModalVisible(true);
    }
  }, [name]);

  const onPressShare = useCallback(async () => {
    const uri = await viewShotRef.current?.capture();
    Share.open({
      title: t('shareTitle', { points: user.points }),
      url: uri,
      type: 'image/jpeg',
    });
  }, [user]);

  //Modal to update name
  const renderModal = () => {
    return (
      <Modal
        transparent={true}
        animationType="slide"
        visible={modalVisible}
        onRequestClose={() => {
          setModalVisible(false);
        }}
      >
        <View style={[Layout.fill, Layout.colHCenter, Common.modalContainer]}>
          <View style={Common.modal}>
            <Text
              style={[Fonts.titleSmall, Fonts.textCenter, Gutters.tinyBMargin]}
            >
              {t('modalTitle')}
            </Text>
            <Text
              style={[Fonts.textRegular, Fonts.textCenter, Gutters.tinyBMargin]}
            >
              {t('modalDesc')}
            </Text>
            <TextInput
              style={[Common.textInput, Fonts.textCenter]}
              ref={nameInputRef}
              value={name || ''}
              onChangeText={setName}
            />
            <Pressable
              disabled={!name || name === ''}
              style={[Common.button.rounded, Gutters.tinyTMargin]}
              onPress={() => {
                console.log(name);
                dispatch(updateUser({ name }));
                setModalVisible(false);
              }}
            >
              <Text style={Fonts.textCenter}>{t('submit')}</Text>
            </Pressable>
          </View>
        </View>
      </Modal>
    );
  };

  return (
    <SafeAreaView style={[Layout.fill, Layout.scrollSpaceAround]}>
      <ViewShot
        style={Layout.scrollSpaceAround}
        ref={viewShotRef}
        options={{ format: 'jpg', quality: 0.9 }}
      >
        <View>
          <Text style={[Fonts.titleRegular, Fonts.textCenter]}>
            {t('title', {
              points,
            })}
          </Text>
          <Text style={[Fonts.titleSmall, Fonts.textCenter]}>
            {t('desc', {
              points: user.points,
            })}
          </Text>
        </View>
        <View>
          <TouchableOpacity
            style={[
              Common.button.rounded,
              Gutters.regularHMargin,
              Gutters.tinyBMargin,
            ]}
            onPress={onPressNext}
          >
            <Text style={Fonts.textCenter}>{t('next')}</Text>
          </TouchableOpacity>
          <TouchableOpacity
            style={[Common.button.rounded, Gutters.regularHMargin]}
            onPress={onPressShare}
          >
            <Text style={Fonts.textCenter}>{t('share')}</Text>
          </TouchableOpacity>
        </View>
      </ViewShot>

      {renderModal()}
    </SafeAreaView>
  );
};

export default Congratz;
