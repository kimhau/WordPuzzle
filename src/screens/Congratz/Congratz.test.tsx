import { fireEvent, render, waitFor } from '@testing-library/react-native';
import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { store } from 'WordPuzzle/src/store';
import Congratz from './Congratz';

const mockStore = configureStore([]);

describe('Congratz', () => {
  let initialStore;
  let component;

  beforeEach(() => {
    initialStore = mockStore({
      user: {
        name: 'John Doe',
        points: 100,
      },
      theme: { theme: 'default', darkMode: null },
    });

    component = (
      <Provider store={initialStore}>
        <Congratz
          route={{
            params: {
              points: 50,
              category: 'Animals',
            },
          }}
          navigation={{ reset: jest.fn() }}
        />
      </Provider>
    );
  });

  it('renders correctly', () => {
    const { getByText, toJSON } = render(component);
    expect(toJSON()).toMatchSnapshot();
    expect(getByText('title')).toBeTruthy();
    expect(getByText('desc')).toBeTruthy();
    expect(getByText('next')).toBeTruthy();
    expect(getByText('share')).toBeTruthy();
  });

  it('calls navigation.reset() when "Next" button is pressed with a name provided', () => {
    const { getByText, getByPlaceholderText } = render(component);

    fireEvent.press(getByText('next'));
    expect(
      component.props.children.props.navigation.reset,
    ).toHaveBeenCalledWith({
      index: 1,
      routes: [
        {
          name: 'Home',
        },
        {
          name: 'Game',
          params: { category: 'Animals' },
        },
      ],
    });
  });

  it('shows a modal when "Next" button is pressed with no name was provided previosly', async () => {
    const { getByText, toJSON, getByRole } = render(
      <Provider store={store}>
        <Congratz
          route={{
            params: {
              points: 50,
              category: 'Animals',
            },
          }}
          navigation={{ reset: jest.fn() }}
        />
      </Provider>,
    );

    fireEvent.press(getByText('next'));
    await waitFor(() => {
      expect(getByText('modalTitle')).toBeTruthy();
      expect(getByText('modalDesc')).toBeTruthy();
      expect(getByText('submit')).toBeTruthy();
    });
  });

  it('calls Share.open() when "Share" button is pressed', async () => {
    const { getByText } = render(component);
    // const openMock = jest.fn();

    const openMock = jest.spyOn(require('react-native-share'), 'open');

    fireEvent.press(getByText('share'));

    await waitFor(() => {
      expect(openMock).toHaveBeenCalledWith({
        title: 'shareTitle',
        url: undefined,
        type: 'image/jpeg',
      });
    });
  });
});
