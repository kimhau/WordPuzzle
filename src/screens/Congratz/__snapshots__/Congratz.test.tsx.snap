// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Congratz renders correctly 1`] = `
<RCTSafeAreaView
  style={
    [
      {
        "flex": 1,
      },
      {
        "flexGrow": 1,
        "justifyContent": "space-around",
      },
    ]
  }
>
  <View>
    <View>
      <Text
        style={
          [
            {
              "color": "#212529",
              "fontSize": 40,
              "fontWeight": "bold",
            },
            {
              "textAlign": "center",
            },
          ]
        }
      >
        title
      </Text>
      <Text
        style={
          [
            {
              "color": "#212529",
              "fontSize": 32,
              "fontWeight": "bold",
            },
            {
              "textAlign": "center",
            },
          ]
        }
      >
        desc
      </Text>
    </View>
    <View>
      <RNGestureHandlerButton
        collapsable={false}
        delayLongPress={600}
        enabled={true}
        exclusive={true}
        handlerTag={1}
        handlerType="NativeViewGestureHandler"
        onGestureEvent={[Function]}
        onGestureHandlerEvent={[Function]}
        onGestureHandlerStateChange={[Function]}
        onHandlerStateChange={[Function]}
        rippleColor={0}
        touchSoundDisabled={false}
      >
        <View
          accessible={true}
          collapsable={false}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#E14032",
              "borderRadius": 10,
              "height": 40,
              "justifyContent": "center",
              "marginBottom": 10,
              "marginHorizontal": 30,
              "opacity": 1,
              "paddingHorizontal": 30,
            }
          }
        >
          <Text
            style={
              {
                "textAlign": "center",
              }
            }
          >
            next
          </Text>
        </View>
      </RNGestureHandlerButton>
      <RNGestureHandlerButton
        collapsable={false}
        delayLongPress={600}
        enabled={true}
        exclusive={true}
        handlerTag={2}
        handlerType="NativeViewGestureHandler"
        onGestureEvent={[Function]}
        onGestureHandlerEvent={[Function]}
        onGestureHandlerStateChange={[Function]}
        onHandlerStateChange={[Function]}
        rippleColor={0}
        touchSoundDisabled={false}
      >
        <View
          accessible={true}
          collapsable={false}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#E14032",
              "borderRadius": 10,
              "height": 40,
              "justifyContent": "center",
              "marginHorizontal": 30,
              "opacity": 1,
              "paddingHorizontal": 30,
            }
          }
        >
          <Text
            style={
              {
                "textAlign": "center",
              }
            }
          >
            share
          </Text>
        </View>
      </RNGestureHandlerButton>
    </View>
  </View>
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={true}
    visible={false}
  />
</RCTSafeAreaView>
`;
